server/aesdsocket.c

```diff
@@ -0,0 +1,242 @@
+	SRC_BIND, SRC_OPEN, SRC_MMAP, SRC_MREMAP, SRC_ACCEPT, SRC_SIGACTION,
+	SRC_MALLOC, SRC_REALLOC, SRC_FSTAT, SRC_CLOSE, SRC_READ, SRC_WRITE,
+	SRC_TERM = SIGTERM, SRC_FTRUNCATE, SRC_EINVAL, SRC_DUP} cleanup_src;
+
+const char* errs[] = {
+	[SRC_INT] = "Caught signal SIGINT, exiting",
+	[SRC_TERM] = "Caught signal SIGTERM, exiting",
+	[SRC_SOCKET] = "error creating server socket",
+	[SRC_LISTEN] = "error listening on server socket",
+	[SRC_BIND] = "error binding server socket",
+	[SRC_OPEN] = "error opening output file",
+	[SRC_MMAP] = "error mapping output file",
+	[SRC_MREMAP] = "error expanding output file mapping",
+	[SRC_ACCEPT] = "error accepting client connection",
+	[SRC_SIGACTION] = "error installing signal handler",
+	[SRC_MALLOC] = "error allocating packet buffer",
+	[SRC_REALLOC] = "error expanding packet buffer",
+	[SRC_FSTAT] = "error in fstat",
+	[SRC_CLOSE] = "error closing client socket",
+	[SRC_READ] = "error reading from client socket",
+	[SRC_WRITE] = "error writing to client socket",
+	[SRC_FTRUNCATE] = "error extending output file",
+	[SRC_EINVAL] = "parsing command line arguments",
+	[SRC_DUP] = "error redirecting stream"
+};
+
+// cleanup handler
+__attribute__((noreturn))
+static void cleanup (cleanup_src src) {
+	// record exit cause
+	const char* msg = errs[src];
+	int xstat = 0;
+	if (src == SRC_INT || src == SRC_TERM) {
+		syslog(LOG_INFO, "%s\n", msg);
+	} else {
+		syslog(LOG_ERR, "%s: %m", msg);
+		xstat = -1;
+	}
+	closelog();
+
+	// clean up state
+	if (csock != -1) close(csock);
+	if (asock != -1) close(asock);
+	if (of_mem) munmap(of_mem, of_memsz);
+	if (ofd != -1) close (ofd);
+	if (packet) free(packet);
+	unlink(outpath);
+
+	exit(xstat);
+}
+
+// wrapper to conform to sigaction struct def
+__attribute__((noreturn))
+static void sigcleanup (int sig) {
+	assert(sig == SIGTERM || sig == SIGINT);
+	cleanup((cleanup_src) sig);
+}
+
+int main (int argc, char** argv) {
+	// open syslog
+	openlog(NULL, LOG_PERROR, LOG_USER);
+
+	// check for daemon arg
+	bool daemon = false;
+	if (argc == 2 && strncmp("-d", argv[1], 3) == 0) daemon = true;
+	else if (argc >= 2) {
+		errno = EINVAL;   
+		cleanup(SRC_EINVAL);
+	}
+	
+	// install signal handler
+	struct sigaction act = {
+		.sa_handler = sigcleanup,
+		.sa_flags = 0
+	};
+	int s = sigaction(SIGINT, &act, NULL);
+	if (s == -1) cleanup(SRC_SIGACTION);
+	s = sigaction(SIGTERM, &act, NULL);
+	if (s == -1) cleanup(SRC_SIGACTION);
+
+	// open+mmap output file
+	const long PAGE_SIZE = sysconf(_SC_PAGESIZE);
+	ofd = open(outpath, O_RDWR | O_CREAT | O_APPEND, 0644);
+	struct stat stat;
+	s = fstat(ofd, &stat);
+	if (s == -1) cleanup(SRC_FSTAT);
+	size_t of_sz = (size_t) stat.st_size;
+	of_memsz = (of_sz/PAGE_SIZE + 1)*PAGE_SIZE;
+	s = ftruncate(ofd, of_memsz);
+	if (s == -1) cleanup(SRC_FTRUNCATE);
+	void* of_mem = mmap(NULL, of_memsz, PROT_READ | PROT_WRITE, MAP_SHARED, ofd, 0);
```

Comment:
@georgehodgkins I'm curious why the memory mapping?  I don't think this is going to work once we make the switch to use your driver instead of the temporaril file in assignment 8.